arangod/Pregel/Algos/HITSKleinberg.cpp

```diff
@@ -0,0 +1,331 @@
+using namespace arangodb;
+using namespace arangodb::pregel;
+
+static std::string const authAggregator = "auth";
+static std::string const hubAggregator = "hub";
+static std::string const maxDiffAuthAggregator = "diffA";
+static std::string const maxDiffHubAggregator = "diffH";
+static std::string const isLastIterationAggregator = "stop";
+
+using VertexType = HITSKleinbergValue;
+
+static double const epsilon = 0.00001;
+
+enum class State {
+  sendInitialHubs,
+  updateAuth,
+  updateAuthNormalizeHub,
+  updateHubNormalizeAuth,
+  finallyNormalizeHubs,
+  finallyNormalizeAuths
+};
+
+struct HITSKleinbergWorkerContext : public WorkerContext {
+  HITSKleinbergWorkerContext(size_t maxGSS, size_t numIterations)
+      : numIterations(numIterations){};
+
+  double authDivisor = 0;
+  double hubDivisor = 0;
+  State state = State::sendInitialHubs;
+  size_t numIterations;
+  size_t currentIteration = 0;
+
+  void preGlobalSuperstep(uint64_t gss) override {
+    // todo: we only need this if the global super-step is odd
+    //  difficulty: the parent class WorkerContext doesn't know the current gss
+    auto const* authNorm = getAggregatedValue<double>(authAggregator);
+    auto const* hubNorm = getAggregatedValue<double>(hubAggregator);
+    authDivisor = std::sqrt(*authNorm);
+    hubDivisor = std::sqrt(*hubNorm);
+
+    double const authMaxDiff =
+        *getAggregatedValue<double>(maxDiffAuthAggregator);
+    double const hubMaxDiff = *getAggregatedValue<double>(maxDiffHubAggregator);
+    const double diff = std::max(authMaxDiff, hubMaxDiff);
+
+    if (diff < epsilon) {
+      if (state == State::updateAuthNormalizeHub) {
+        state = State::finallyNormalizeHubs;
+      } else if (state == State::updateHubNormalizeAuth) {
+        state = State::finallyNormalizeAuths;
+      }
+    }
+  }
+
+  void postGlobalSuperstep(uint64_t gss) override {
+    switch (state) {
+      case State::sendInitialHubs:
+        state = State::updateAuth;
+        break;
+      case State::updateAuth:
+      case State::updateAuthNormalizeHub:
+        state = State::updateHubNormalizeAuth;
+        break;
+      case State::updateHubNormalizeAuth:
+        ++currentIteration;
+        if (currentIteration == numIterations) {
+          state = State::finallyNormalizeHubs;
+        } else {
+          state = State::updateAuthNormalizeHub;
+        }
+        break;
+      case State::finallyNormalizeHubs:
+      case State::finallyNormalizeAuths:
+        aggregate(isLastIterationAggregator, true);
+        break;
+    }
+  }
+};
+
+struct HITSKleinbergComputation
+    : public VertexComputation<VertexType, int8_t, SenderMessage<double>> {
+  HITSKleinbergComputation() = default;
+
+  void sendAuthToInNeighbors(
+      MessageIterator<SenderMessage<double>> const& receivedMessages,
+      double auth) {
+    SenderMessage<double> authData(this->pregelId(), auth);
+    for (SenderMessage<double> const* message : receivedMessages) {
```

Comment:
Is there a message sent to every neighbour in each GSS otherwise you might miss some neighbours if they didn't talk to you.

Reply:
Yes, there is. I added a corresponding comment.